src/main/java/com/salesforce/dataloader/security/EncryptionAesUtil.java

```diff
@@ -0,0 +1,255 @@
+            } else if (OS.indexOf("win") >= 0) {
+                detectedOS = OSType.Windows;
+            } else if (OS.indexOf("nux") >= 0) {
+                detectedOS = OSType.Linux;
+            } else {
+                detectedOS = OSType.Other;
+            }
+        }
+        return detectedOS;
+    }
+
+    // Support single text encryption and decryption
+
+    static public Cipher cipher;
+    static public byte[] cipherKey;
+
+    // 16 bytes was used for 128 bit AES encryption
+    public static final int ENCRYPTION_KEY_LENGTH_IN_BYTES = 16;
+    public static final String DEFAULT_KEYFILE_NAME = "dataLoader.key";
+
+    static {
+        try {
+            cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
+            detectedOS = getOperatingSystemType();
+
+        } catch (Exception e) {
+            LOGGER.error("Fail to initialize encryption: " + e.getMessage());
+            throw new RuntimeException("Fail to initialize encryption:  ", e);
+        }
+    }
+
+
+    private byte[] extractIvBytes(byte[] cipheredText) {
+        byte[] iv = new byte[ENCRYPTION_KEY_LENGTH_IN_BYTES];
+        System.arraycopy(cipheredText, 0, iv, 0, ENCRYPTION_KEY_LENGTH_IN_BYTES);
+        return iv;
+    }
+
+    private byte[] extractCipherContent(byte[] cipheredText) {
+        byte[] cipheredPassword = new byte[cipheredText.length - ENCRYPTION_KEY_LENGTH_IN_BYTES];
+        System.arraycopy(cipheredText, ENCRYPTION_KEY_LENGTH_IN_BYTES, cipheredPassword, 0, cipheredText.length - ENCRYPTION_KEY_LENGTH_IN_BYTES);
+        return cipheredPassword;
+
+    }
+
+    private byte[] concatenateByteArray(byte[] a, byte[] b) {
+
+        byte[] c = new byte[a.length + b.length];
+        System.arraycopy(a, 0, c, 0, a.length);
+        System.arraycopy(b, 0, c, a.length, b.length);
+        return c;
+    }
+
+    public byte[] generateEncryptionKey() {
+        // The secure random uses threadId and tick to make sure it is not repeatable
+        // If a call to setSeed had not occurred previously,
+        // the first call to this method forces this SecureRandom object to seed itself.
+        // This self-seeding will not occur if setSeed was previously called.
+        SecureRandom random = new SecureRandom();
+        byte bytes[] = new byte[ENCRYPTION_KEY_LENGTH_IN_BYTES];
+        random.nextBytes(bytes);
+        return bytes;
+    }
+
+    public String createUserProfileKeyName() {
+        String path = Paths.get(System.getProperty("user.home"), ".dataloader").toString();
+        File customDir = new File(path);
+
+        if (customDir.exists()) {
+            LOGGER.debug(customDir + " exists");
+        } else if (customDir.mkdirs()) {
+            if (detectedOS == OSType.MacOS || detectedOS == OSType.Linux) {
+                // set all reading to false
+                customDir.setReadable(false, false);
+                // only owner can read
+                customDir.setReadable(true, true);
```

Comment:
You should remove execute permissions (from everybody other than the owner) from the containing directory as well, as an extra layer of protection for files inside (dirs can't be listed without Read permissions but can be traversed to access their contents - assuming the names are known - unless Execute permissions are also removed).